File: dsl/java-test/java-client/src/test/java/dev/rsdlang/sample/client/ScalarSubstition_ServiceServiceTest.java
package dev.rsdlang.sample.client;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;

import java.net.URI;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import dev.rsdlang.sample.client.jdkhttp.JDKSpecSamplesClient;
import dev.rsdlang.sample.client.jdkhttp.JDKSpecSamplesClient.ContentTypeEncoding;
import dev.rsdlang.sample.client.model.NilResult;
import dev.rsdlang.sample.client.model.ZoneId;

public class ScalarSubstition_ServiceServiceTest {
    private static SpecSamplesClient JSON;
    private static SpecSamplesClient MSGPACK;

    @BeforeAll
    static void setUp() {
        var baseBuilder = JDKSpecSamplesClient.builder().baseURI(URI.create("http://localhost:3000"));
        JSON = baseBuilder.build();
        MSGPACK = baseBuilder.contentTypeEncoding(ContentTypeEncoding.APPLICATION_VND_MSGPACK).build();
    }

    static ScalarSubstition_ServiceService[] serviceProvider() {
        return new ScalarSubstition_ServiceService[] {
                JSON.service(ScalarSubstition_ServiceService.class),
                MSGPACK.service(ScalarSubstition_ServiceService.class),
        };
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void get(ScalarSubstition_ServiceService service) {
        var result = service.get().orThrow();
        assertEquals("[0,0]", MyRange.toString(result));
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void list(ScalarSubstition_ServiceService service) {
        var result = service.list().orThrow();
        assertEquals(1, result.size());
        assertEquals("[0,0]", MyRange.toString(result.get(0)));
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void post(ScalarSubstition_ServiceService service) {
        var result = service.post(new MyRange(1, 1)).orThrow();
        assertEquals("[1,1]", MyRange.toString(result));
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void postOpt(ScalarSubstition_ServiceService service) {
        assertEquals(NilResult.UNDEFINED, service.postOpt().orThrow());
        assertEquals(NilResult.DEFINED, service.postOpt(new MyRange(1, 1)).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void postNull(ScalarSubstition_ServiceService service) {
        assertEquals(NilResult.NULL, service.postNull(null).orThrow());
        assertEquals(NilResult.DEFINED, service.postNull(new MyRange(1, 1)).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void postOptNull(ScalarSubstition_ServiceService service) {
        assertEquals(NilResult.UNDEFINED, service.postOptNull().orThrow());
        assertEquals(NilResult.NULL, service.postOptNull(null).orThrow());
        assertEquals(NilResult.DEFINED, service.postOptNull(new MyRange(1, 1)).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void postList(ScalarSubstition_ServiceService service) {
        var result = service.postList(List.of(new MyRange(1, 1))).orThrow();
        assertEquals(1, result.size());
        assertEquals("[1,1]", MyRange.toString(result.get(0)));
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void postListOpt(ScalarSubstition_ServiceService service) {
        assertEquals(NilResult.UNDEFINED, service.postListOpt().orThrow());
        assertEquals(NilResult.DEFINED, service.postListOpt(List.of(new MyRange(1, 1))).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void postListNull(ScalarSubstition_ServiceService service) {
        assertEquals(NilResult.NULL, service.postListNull(null).orThrow());
        assertEquals(NilResult.DEFINED, service.postListNull(List.of(new MyRange(0, 0))).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void postListOptNull(ScalarSubstition_ServiceService service) {
        assertEquals(NilResult.UNDEFINED, service.postListOptNull().orThrow());
        assertEquals(NilResult.NULL, service.postListOptNull(null).orThrow());
        assertEquals(NilResult.DEFINED, service.postListOptNull(List.of(new MyRange(0, 0))).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void query(ScalarSubstition_ServiceService service) {
        var result = service.query(new MyRange(1, 1)).orThrow();
        assertEquals("[1,1]", MyRange.toString(result));
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void queryOpt(ScalarSubstition_ServiceService service) {
        assertEquals(NilResult.UNDEFINED, service.queryOpt().orThrow());
        assertEquals(NilResult.DEFINED, service.queryOpt(new MyRange(1, 1)).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void queryNull(ScalarSubstition_ServiceService service) {
        assertEquals(NilResult.NULL, service.queryNull(null).orThrow());
        assertEquals(NilResult.DEFINED, service.queryNull(new MyRange(1, 1)).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void queryOptNull(ScalarSubstition_ServiceService service) {
        assertEquals(NilResult.UNDEFINED, service.queryOptNull().orThrow());
        assertEquals(NilResult.NULL, service.queryOptNull(null).orThrow());
        assertEquals(NilResult.DEFINED, service.queryOptNull(new MyRange(1, 1)).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void queryList(ScalarSubstition_ServiceService service) {
        var result = service.queryList(List.of(new MyRange(1, 1))).orThrow();
        assertEquals(1, result.size());
        assertEquals("[1,1]", MyRange.toString(result.get(0)));
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void queryListOpt(ScalarSubstition_ServiceService service) {
        // assertEquals(NilResult.UNDEFINED, service.queryListOpt());
        assertEquals(NilResult.DEFINED, service.queryListOpt(List.of(new MyRange(1,
                1))).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void queryListNull(ScalarSubstition_ServiceService service) {
        // assertEquals(NilResult.NULL, service.queryListNull(null));
        assertEquals(NilResult.DEFINED, service.queryListNull(List.of(new MyRange(1, 1))).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void queryListOptNull(ScalarSubstition_ServiceService service) {
        // assertEquals(NilResult.UNDEFINED, service.queryListOptNull());
        // assertEquals(NilResult.NULL, service.queryListOptNull(null));
        assertEquals(NilResult.DEFINED, service.queryListOptNull(List.of(new MyRange(1, 1))).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void header(ScalarSubstition_ServiceService service) {
        var result = service.header(new MyRange(1, 1)).orThrow();
        assertEquals("[1,1]", MyRange.toString(result));
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void headerOpt(ScalarSubstition_ServiceService service) {
        assertEquals(NilResult.UNDEFINED, service.headerOpt().orThrow());
        assertEquals(NilResult.DEFINED, service.headerOpt(new MyRange(1, 1)).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void headerNull(ScalarSubstition_ServiceService service) {
        assertEquals(NilResult.NULL, service.headerNull(null).orThrow());
        assertEquals(NilResult.DEFINED, service.headerNull(new MyRange(1, 1)).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void headerOptNull(ScalarSubstition_ServiceService service) {
        assertEquals(NilResult.UNDEFINED, service.headerOptNull().orThrow());
        assertEquals(NilResult.NULL, service.headerOptNull(null).orThrow());
        assertEquals(NilResult.DEFINED, service.headerOptNull(new MyRange(1, 1)).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void headerList(ScalarSubstition_ServiceService service) {
        var result = service.headerList(List.of(new MyRange(0, 0))).orThrow();
        assertEquals(1, result.size());
        assertEquals("[0,0]", MyRange.toString(result.get(0)));
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void headerListOpt(ScalarSubstition_ServiceService service) {
        assertEquals(NilResult.UNDEFINED, service.headerListOpt().orThrow());
        assertEquals(NilResult.DEFINED, service.headerListOpt(List.of(new MyRange(0, 0))).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void headerListNull(ScalarSubstition_ServiceService service) {
        assertEquals(NilResult.NULL, service.headerListNull(null).orThrow());
        assertEquals(NilResult.DEFINED, service.headerListNull(List.of(new MyRange(0, 0))).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void headerListOptNull(ScalarSubstition_ServiceService service) {
        assertEquals(NilResult.UNDEFINED, service.headerListOptNull().orThrow());
        assertEquals(NilResult.NULL, service.headerListOptNull(null).orThrow());
        assertEquals(NilResult.DEFINED, service.headerListOptNull(List.of(new MyRange(0, 0))).orThrow());
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void fail(ScalarSubstition_ServiceService service) {
        var result = service.fail();
        switch (result) {
            case Result.OK(var value) -> Assertions.fail("Expected SampleErrorScalarSubException to be thrown");
            case Result.ERR(SampleErrorScalarSub error) -> {
                assertEquals("This is a sample error from the server", error.message());
                assertEquals("[0,0]", MyRange.toString(error.data()));
            }
            default -> Assertions.fail("Unexpected result type: " + result.getClass().getName());
        }
    }

    @ParameterizedTest
    @MethodSource("serviceProvider")
    public void multiBody(ScalarSubstition_ServiceService service) {
        var result = service.multiBody(new MyRange(1, 1), ZoneId.of("Europe/Berlin")).orThrow();
        assertEquals("[1,1] Europe/Berlin", result);
    }
}

File: dsl/packages/cli/test/generated-code/typescript-rest-client-fetch/ScalarSubstition_ServiceService.test.ts
import { describe, expect, test } from 'vitest';
import { createScalarSubstition_ServiceService } from '../../test-specs/gen-out/client/typescript-client/src/index.js';
import { isSampleErrorScalarSubError } from '../../test-specs/gen-out/client/typescript-client/src/Errors.js';

const jsonService = createScalarSubstition_ServiceService({
	baseUrl: 'http://localhost:3000',
});

const msgpackService = createScalarSubstition_ServiceService({
	baseUrl: 'http://localhost:3000',
	encoding: 'application/vnd.msgpack',
});

/*const openApiService = createOpenAPISampleServiceService({
	baseUrl: 'http://localhost:3000',
});
*/

const json = {
	encoding: 'application/json' as const,
	service: jsonService,
};

const msgpack = {
	encoding: 'application/vnd.msgpack' as const,
	service: msgpackService,
};

describe('ScalarSubstition_ServiceService', () => {
	describe('get', () => {
		test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
			const [result, error] = await service.get();
			expect(error).toBeNull();
			expect(result).toBeDefined();
			expect(result?.start).toBe(0);
			expect(result?.end).toBe(0);
		});
	});
	describe('list', () => {
		test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
			const [result, error] = await service.list();
			expect(error).toBeNull();
			expect(result).toBeDefined();
			expect(result?.length).toBe(1);
			expect(result?.[0].start).toBe(0);
			expect(result?.[0].end).toBe(0);
		});
	});
	describe('post', () => {
		test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
			const [result, error] = await service.post({ start: 1, end: 2 });
			expect(error).toBeNull();
			expect(result).toBeDefined();
			expect(result?.start).toBe(1);
			expect(result?.end).toBe(2);
		});
	});
	describe('postOpt', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.postOpt({ start: 1, end: 2 });
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		describe('undefined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.postOpt();
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
	});
	describe('postNull', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.postNull({ start: 1, end: 2 });
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		describe('null', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.postNull(null);
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
	});
	describe('postOptNull', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.postOptNull({ start: 1, end: 2 });
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		describe('null', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.postOptNull(null);
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		describe('undefined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.postOptNull();
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
	});
	describe('postList', () => {
		test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
			const [result, error] = await service.postList([{ start: 1, end: 2 }]);
			expect(error).toBeNull();
			expect(result).toBeDefined();
			expect(result?.length).toBe(1);
			expect(result?.[0].start).toBe(1);
			expect(result?.[0].end).toBe(2);
		});
	});
	describe('postListOpt', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.postListOpt([{ start: 1, end: 2 }]);
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		describe('undefined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.postListOpt();
				expect(error).toBeNull();
				expect(result).toBe('UNDEFINED');
			});
		});
	});
	describe('postListNull', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.postListNull([{ start: 1, end: 2 }]);
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		describe('null', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.postListNull(null);
				expect(error).toBeNull();
				expect(result).toBe('NULL');
			});
		});
	});
	describe('postListOptNull', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.postListOptNull([{ start: 1, end: 2 }]);
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		describe('null', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.postListOptNull(null);
				expect(error).toBeNull();
				expect(result).toBe('NULL');
			});
		});
		describe('undefined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.postListOptNull();
				expect(error).toBeNull();
				expect(result).toBe('UNDEFINED');
			});
		});
	});
	describe('query', () => {
		test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
			const [result, error] = await service.query({ start: 1, end: 2 });
			expect(error).toBeNull();
			expect(result).toBeDefined();
			expect(result?.start).toBe(1);
			expect(result?.end).toBe(2);
		});
	});
	describe('queryOpt', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.queryOpt({ start: 1, end: 2 });
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		describe('undefined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.queryOpt();
				expect(error).toBeNull();
				expect(result).toBe('UNDEFINED');
			});
		});
	});
	describe('queryNull', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.queryNull({ start: 1, end: 2 });
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		/*describe('null', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.queryNull(null);
				expect(error).toBeNull();
				expect(result).toBe('NULL');
			});
		});*/
	});
	describe('queryOptNull', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.queryOptNull({ start: 1, end: 2 });
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		/*escribe('null', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.queryOptNull(null);
				expect(error).toBeNull();
				expect(result).toBe('NULL');
			});
		});*/
		describe('undefined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.queryOptNull();
				expect(error).toBeNull();
				expect(result).toBe('UNDEFINED');
			});
		});
	});
	describe('queryList', () => {
		test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
			const [result, error] = await service.queryList([{ start: 1, end: 2 }]);
			expect(error).toBeNull();
			expect(result).toBeDefined();
			expect(result?.length).toBe(1);
			expect(result?.[0].start).toBe(1);
			expect(result?.[0].end).toBe(2);
		});
	});
	describe('queryListOpt', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.queryListOpt([{ start: 1, end: 2 }]);
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		/*describe('undefined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.queryListOpt();
				expect(error).toBeNull();
				expect(result).toBe('UNDEFINED');
			});
		});*/
	});
	describe('queryListNull', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.queryListNull([{ start: 1, end: 2 }]);
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		/*describe('null', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.queryListNull(null);
				expect(error).toBeNull();
				expect(result).toBe('NULL');
			});
		});*/
	});
	describe('queryListOptNull', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.queryListOptNull([{ start: 1, end: 2 }]);
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		/*describe('undefined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.queryListOptNull();
				expect(error).toBeNull();
				expect(result).toBe('UNDEFINED');
			});
		});
		describe('null', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.queryListOptNull(null);
				expect(error).toBeNull();
				expect(result).toBe('NULL');
			});
		});*/
	});
	describe('header', () => {
		test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
			const [result, error] = await service.header({ start: 1, end: 2 });
			expect(error).toBeNull();
			expect(result).toBeDefined();
			expect(result?.start).toBe(1);
			expect(result?.end).toBe(2);
		});
	});
	describe('headerOpt', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.headerOpt({ start: 1, end: 2 });
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		describe('undefined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.headerOpt();
				expect(error).toBeNull();
				expect(result).toBe('UNDEFINED');
			});
		});
	});
	describe('headerNull', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.headerNull({ start: 1, end: 2 });
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		describe('null', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.headerNull(null);
				expect(error).toBeNull();
				expect(result).toBe('NULL');
			});
		});
	});
	describe('headerOptNull', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.headerOptNull({ start: 1, end: 2 });
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		describe('null', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.headerOptNull(null);
				expect(error).toBeNull();
				expect(result).toBe('NULL');
			});
		});
		describe('undefined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.headerOptNull();
				expect(error).toBeNull();
				expect(result).toBe('UNDEFINED');
			});
		});
	});
	describe('headerList', () => {
		test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
			const [result, error] = await service.headerList([{ start: 1, end: 2 }]);
			expect(error).toBeNull();
			expect(result).toBeDefined();
			expect(result?.length).toBe(1);
			expect(result?.[0].start).toBe(1);
			expect(result?.[0].end).toBe(2);
		});
	});
	describe('headerListOpt', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.headerListOpt([{ start: 1, end: 2 }]);
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		describe('undefined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.headerListOpt();
				expect(error).toBeNull();
				expect(result).toBe('UNDEFINED');
			});
		});
	});
	describe('headerListNull', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.headerListNull([{ start: 1, end: 2 }]);
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		describe('null', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.headerListNull(null);
				expect(error).toBeNull();
				expect(result).toBe('NULL');
			});
		});
	});
	describe('headerListOptNull', () => {
		describe('defined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.headerListOptNull([{ start: 1, end: 2 }]);
				expect(error).toBeNull();
				expect(result).toBeDefined();
			});
		});
		describe('null', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.headerListOptNull(null);
				expect(error).toBeNull();
				expect(result).toBe('NULL');
			});
		});
		describe('undefined', () => {
			test.each([json, msgpack])('ssucess with $encoding', async ({ service }) => {
				const [result, error] = await service.headerListOptNull();
				expect(error).toBeNull();
				expect(result).toBe('UNDEFINED');
			});
		});
	});
	describe('fail', () => {
		test.each([json, msgpack])('fail with $encoding', async ({ service }) => {
			const [result, error] = await service.fail();
			expect(result).toBeUndefined();
			expect(error).toBeDefined();
			if (isSampleErrorScalarSubError(error)) {
				// expect(error.data).toBe();
			}
		});
	});
});
